### PR TITLE
Delay building web.browser.legacy bundle until after development server restarts.

### PR DESCRIPTION
**Pull request by benjamn**
Now that Meteor [builds two client bundles](https://blog.meteor.com/meteor-1-7-and-the-evergreen-dream-a8c1270b0901), build times have naturally increased a bit. We are hard at work attacking this problem from multiple angles, such as #9983, and I'm confident that Meteor 1.7.1 build times will be shorter for most apps than they were before Meteor 1.7, despite building both `web.browser` and `web.browser.legacy` bundles.

Nevertheless, since most testing in development happens using one bundle/browser at a time, it has been suggested (#9948) that we could skip the legacy build in development, as a cheap short-cut to faster rebuild times. To be honest, I don't love most of these ideas for disabling the legacy build with some sort of additional configuration, because supporting older browsers is important, configuration is a creeping cancer, and ignoring errors in the legacy build seems like a recipe for frustration when it's time to deploy your app.

I would vastly prefer a zero-configuration solution to this problem, though that's challenging in this case because skipping an important part of the build process seems like a decision the developer should make consciously and carefully.

As this pull request demonstrates, it's possible to remove the legacy build from the critical path to restarting the development server, so you can continue testing and developing your app while the legacy bundle is built in the background. There's nothing to configure, because the legacy bundle still gets built in roughly the same amount of time as before; you just don't have to wait for it to finish, if you aren't currently interested in testing legacy code.

How it works in detail:
* All bundles are built when the build tool first starts up, so that there is always a legacy bundle available, even if it's slightly stale, and commands like `meteor build` and `meteor deploy` still work as before.
* On each subsequent rebuild of the application, we delay starting the `web.browser.legacy` build until after all other builds have finished and the development server has restarted.
* If you access your application using a legacy browser before the legacy build has finished, you'll just get the previous (possibly slightly stale) legacy code. Not the end of the world!
* As soon as the legacy build finishes, the build process sends a message to the `webapp` package running in the server process to tell it to reload the latest legacy bundle.
* Any errors building the legacy bundle get reported in the same way as before.
* Once everything has finished, you'll see the following message in the console:
  ```
  => Finished delayed build of web.browser.legacy in 951ms
  ```
  This message means it's now safe to test your app in a legacy browser, if that's something you were trying to do.

In short, when we talk about zero-configuration here at the Meteor project, we mean something much  deeper than just picking sensible defaults. No, we are striving to eliminate the _need_ for configuration whenever and wherever we can. If you want to hear more about this cornerstone of the Meteor philosophy, check out this [Meteor Night talk](https://youtu.be/vpCotlPieIY?t=34m35s) that I recently gave about the Meteor 1.7 modern/legacy bundling system (or read the [slides](https://slides.com/benjamn/meteor-night-may-2018#/49)).

### CODE REVIEW

File: packages/webapp/webapp_server.js [hunk 1]
```diff
@@ -571,92 +567,114 @@ function runWebAppServer() {
...
-              manifest,
-              null,
-              _.pick(__meteor_runtime_config__, "PUBLIC_SETTINGS")
-            ),
-          cordovaCompatibilityVersions: clientJson.cordovaCompatibilityVersions,
-          PUBLIC_SETTINGS: __meteor_runtime_config__.PUBLIC_SETTINGS
-        };
+  process.on("message", async (message) => {
+    if (message.package !== "webapp") return;
+    if (message.method === "generateClientProgram") {
```
Comment (benjamn) [inline packages/webapp/webapp_server.js]: It's possible that the parent process might send this message before this handler has been registered, but that's ok, because the changes will get picked up when we call `WebAppInternals.reloadClientPrograms` for the first time (which happens later).

**Comment by hwillson** [comment]
This is awesome @benjamn! One small thing:

1. `meteor create test-app`
2. Started the app and accessed it via http://localhost:3100/. So far so good.
3. Added a `console.log('test');` to `client/main.js`.
4. Saw the proper rebuild messages in the console:
```sh
=> Client modified -- refreshing
=> Finished delayed build of web.browser.legacy in 233ms
```
5. Noticed that http://localhost:3100/ did not autoupdate; the newly added `console.log` didn't run.
6. If I manually refresh the browser, it then shows.

**Comment by benjamn** [comment]
@hwillson Great find! I've tracked this down to the code in `webapp` where we calculate client hashes:
https://github.com/meteor/meteor/blob/265d9e30d4f7147e69283f4759bcc8c1794ea226/packages/webapp/webapp_server.js#L191-L209
The problem is that `WebApp.defaultArch` is exactly the architecture that we delayed rebuilding:
https://github.com/meteor/meteor/blob/265d9e30d4f7147e69283f4759bcc8c1794ea226/packages/webapp/webapp_server.js#L42-L44
Since the client hash is the same for `web.browser.legacy` as before, we don't trigger the autoupdate. Seems like this hashing logic should be updated to consider all client bundles, not just `WebApp.defaultArch`.

**Comment by benjamn** [comment]
Although tests are passing after my rewrite of the `autoupdate` package, I'm still concerned about the lack of coordination between the `webapp` package reading the contents of the `.meteor/local/build/programs/web.browser.legacy` bundle during server startup (in the child process), and the build process rewriting the contents of the `web.browser.legacy` bundle (also during server startup, but from the parent process).

I can make the delayed writing of the legacy bundle "more atomic" by populating a fresh temporary directory and then renaming that directory, but that's slower overall because it means rewriting everything rather than just the files that have changed. Writing the changed files in place (and deleting any files that were removed) is fundamentally less instantaneous than renaming a directory, but _much_ faster because it doesn't require creating a temporary directory from scratch.

Here are some ideas I'm considering to address this read/write race:
* Temporarily rename the whole existing `.meteor/local/build/programs/web.browser.legacy` directory to a temporary location, write any changed files, and then rename it back. The `webapp` package could then skip loading the legacy bundle if the directory is missing, trusting that it will be loaded later once the bundle has been written. Problems with this approach:
  * The `webapp` package could start reading the legacy bundle before the build process has a chance to rename it, and I'm not exactly sure how to make the build process wait for the `webapp` package to finish reading the bundle.

* Start writing the `web.browser.legacy` build even later, after the `webapp` package has finished reading the (stale) bundle, so that there's no overlap. The good news here is that we only need to delay the writing of the bundle, but the in-memory computation of the bundle contents (`makeClientTarget`) can begin as early as we like. Problems:
  * The legacy build finishes even later.

* Use some sort of file system synchronization such as [`lockfile`s](https://www.npmjs.com/package/lockfile). Problems:
  * This package (and related packages) appear to be based on the [`O_EXCL` flag](https://nodejs.org/api/fs.html#fs_file_system_flags), which doesn't work on all file systems (e.g. network file systems like NTFS).

I'm going to keep thinking about this until I have more confidence in one solution over the others.

File: tools/isobuild/bundler.js [hunk 2]
```diff
@@ -3181,80 +3138,92 @@ function bundle({
...
+          writeClientTarget(target);
+
+          // Refresh and unpause the client, now that writing is finished.
+          await refreshClient(arch);
+
+          // Let the webapp package running in the child process know it
+          // should regenerate the client program for this arch.
+          runLog.log(`Finished delayed build of ${arch} in ${
+            new Date - start
+          }ms`, { arrow: true });
```
Comment (benjamn) [inline tools/isobuild/bundler.js]: Need to think about whether this logging should remain in the final version. Alternatively, we could add more information about building other bundles, so that this isn't the only message.

File: packages/meteor/message-dispatch.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    payload,
+  }) => {
+    if (type === "FROM_PARENT" &&
+        typeof packageName === "string") {
+      // Keep the messages and their responses strictly ordered per
+      // package, one after the last. The first message waits for the
+      // package to call Package._define(packageName, exports).
+      promises[packageName] = (
+        promises[packageName] || new Promise(resolve => {
+          Package._on(packageName, resolve);
```
Comment (benjamn) [inline packages/meteor/message-dispatch.js]: This pattern causes a bug when the `autoupdate` package (or any other package that receives messages) is not installed, which means `Package._on("autoupdate", resolve)` will never call `resolve`, so the build process stops receiving messages from the server process and `await`s forever.
